Bug 1168134 - URL Bar should color the domain differently from the rest of the URL

sleroux opened the pull request.

thebnich commented on Client/Frontend/Browser/BrowserLocationView.swift:
```diff
@@ -132,13 +138,23 @@ class BrowserLocationView : UIView, UIGestureRecognizerDelegate {
-                    locationLabel.text = url.substringFromIndex(advance(url.startIndex, 8))
+
+            if let urlString = url?.absoluteString {
+
```
Nit: drop following newline for consistency with our other `if` statements

thebnich commented on Client/Frontend/Browser/BrowserLocationView.swift:
```diff
@@ -132,13 +138,23 @@ class BrowserLocationView : UIView, UIGestureRecognizerDelegate {
+
+            if let urlString = url?.absoluteString {
+
+                var truncatedURL: String
```
I think Swift now allows you to use `let` here (since it's guaranteed to be set once the `if` statement is done).

thebnich commented on Client/Frontend/Browser/BrowserLocationView.swift:
```diff
@@ -132,13 +138,23 @@ class BrowserLocationView : UIView, UIGestureRecognizerDelegate {
+
+                var truncatedURL: String
+                if let scheme = url?.scheme {
+                    truncatedURL = urlString.substringFromIndex(advance(urlString.startIndex, count(scheme) + count("://")))
```
I think we may want to stick with explicitly checking `http://` here rather than removing all schemes. If a page uses HTTPS, we want it to be visible in the URL bar so it's extra clear to the user (at least, this is what we do on Firefox Android and desktop).

Also, I think it's possible to have a URL with a scheme that doesn't have `://` in the string (e.g., `about:home`).

sleroux replied:
Cool - I know I before we were stripping out the https:// but I can change it so that we only strip out http:// but keep https:// to show the user we're displaying a https site

sleroux replied:
I also just noticed that on desktop we don't highlight the www. if it's visible where as in this branch we do. I'll update it to match desktop/android.

wesj replied:
Desktop (and Android's) behavior is a bit nuts, but the point of this is to not highlight ANY subdomains (since they allow for easy spoofing i.e. www.google.com.realsite.com). There's a huge list of tld's here that desktop and Android use for that: https://publicsuffix.org/list/effective_tld_names.dat

sleroux commented:
So I updated the patch to include changes in the PR comments and refactored out the logic for coloring the URL into extension methods on NSURL and NSAttributedString so I could wrap some of the behavior in tests. To color the host without the subdomains, I'm currently using the last two components of the host, joining them, and coloring that in hopes of excluding any subdomains. Would this work as a workaround? If not, I can look into using the tld.dat list to create special handling and wrap that in some additional tests if needed.

sleroux commented:
I managed to find http://mxr.mozilla.org/mozilla-central/source/netwerk/dns/nsEffectiveTLDService.cpp#224 in Gecko which is the code thats responsible for determining the public suffix/base domain for a given hostname using the public suffix list found here: https://publicsuffix.org/. I implemented the algorithm in Swift as an extension to NSURL so we can match the domain highlighting logic we have on Android/Desktop. I also included some basic test cases to confirm the algorithm's logic.

sleroux commented:
lttm

fxios-buildasaur commented:
Result of integration 1
Integration took 12 minutes and 3 seconds.
All 130 tests passed, but please fix 17 warnings.

fxios-buildasaur commented:
Result of integration 2
Integration took 12 minutes and 34 seconds.
All 130 tests passed, but please fix 17 warnings.
